pgxn/neon/pagestore_smgr.c

```diff
@@ -126,35 +127,335 @@ static UnloggedBuildPhase unlogged_build_phase = UNLOGGED_BUILD_NOT_IN_PROGRESS;
+ * data to the smgr.
+ */
+void
+prefetch_on_ps_disconnect(void)
+{
+	if (inner_ps_disconnect_hook != NULL)
+		inner_ps_disconnect_hook();
+
+	for (; MyPState->ring_receive < MyPState->ring_unused; MyPState->ring_receive++)
+	{
+		PrefetchRequest *slot;
+		int		index = MyPState->ring_receive % READ_BUFFER_SIZE;
+
+		slot = &MyPState->prf_buffer[index];
+		Assert(slot->status == PRFS_REQUESTED);
+		Assert(slot->my_ring_index == MyPState->ring_receive);
+
+		/* clean up the request */
+		slot->status = PRFS_TAG_REMAINS;
+		prefetch_set_unused(MyPState->ring_receive);
+	}
+}
+
+/*
+ * prefetch_set_unused() - clear a received prefetch slot
+ *
+ * The slot at ring_index must be a current member of the ring buffer,
+ * and may not be in the PRFS_REQUESTED state.
+ */
+static inline void
+prefetch_set_unused(uint64 ring_index)
+{
+	PrefetchRequest *slot = &MyPState->prf_buffer[ring_index % READ_BUFFER_SIZE];
+
+	if (slot->status == PRFS_UNUSED)
+		return;
+
+	Assert(slot->status == PRFS_RECEIVED || slot->status == PRFS_TAG_REMAINS);
+	Assert(ring_index >= MyPState->ring_last &&
+		   ring_index < MyPState->ring_unused);
+
+	if (slot->status == PRFS_RECEIVED)
+	{
+		pfree(slot->response);
+		slot->response = NULL;
+
+		MyPState->n_responses_buffered -= 1;
+		MyPState->n_unused += 1;
+	}
+	else
+	{
+		Assert(slot->response == NULL);
+	}
+
+	prfh_delete(MyPState->prf_hash, slot);
+
+	/* clear all fields */
+	MemSet(slot, 0, sizeof(PrefetchRequest));
+	slot->status = PRFS_UNUSED;
+
+	/* run cleanup if we're holding back ring_last */
+	if (MyPState->ring_last == ring_index)
+		prefetch_cleanup();
+}
+
+static void
+prefetch_do_request_ext(PrefetchRequest *slot, bool *force_latest, XLogRecPtr *force_lsn)
+{
+	NeonGetPageRequest request = {
+		.req.tag = T_NeonGetPageRequest,
+		.req.latest = false,
+		.req.lsn = 0,
+		.rnode = slot->buftag.rnode,
+		.forknum = slot->buftag.forkNum,
+		.blkno = slot->buftag.blockNum,
+	};
+	slot->effective_request_lsn = prefetch_lsn;
+	
+	if (force_lsn && force_latest)
+	{
+		request.req.lsn = *force_lsn;
+		request.req.latest = *force_latest;
+	}
+	else
+	{
+		request.req.lsn = neon_get_request_lsn(
+			&request.req.latest,
+			slot->buftag.rnode,
+			slot->buftag.forkNum,
+			slot->buftag.blockNum
+		);
+	}
+
+	MyPState->n_requestes_inflight++;
+
+	page_server->send((NeonRequest *) &request);
+	Assert(slot->response == NULL);
+	slot->status = PRFS_REQUESTED;
+}
+
+/*
+ * prefetch_register_buffer() - register and prefetch buffer
+ *
+ * Register that we may want the contents of BufferTag in the near future.
+ */
+static uint64
+prefetch_register_buffer(BufferTag tag)
+{
+	return prefetch_register_buffer_ext(tag, NULL, NULL);
+}
+
+static uint64
+prefetch_register_buffer_ext(BufferTag tag, bool *force_latest, XLogRecPtr *force_lsn)
+{
+	int		index;
+	bool	found;
+	uint64	ring_index;
+	PrefetchRequest *slot;
+	PrfHashEntry *entry;
+
+	entry = prfh_lookup(MyPState->prf_hash, (PrefetchRequest *) &tag);
```

Comment:
This function does a `prfh_lookup`, and later `prfh_insert`. You could save some cycles by doing `prfh_insert` directly, and checking the return code for whether it already existed. I'm not sure if there would be problems if the `prefetch_set_unused` calls below remove entries from the hash table, I don't remember if it's safe to keep accessing an entry returned by `prfh_insert`, if you modify the hash table at the same time.

If we don't dare to do that, an easier win would be calculate the hash key only once, and call `prfh_lookup_hash` and `prfh_insert_hash`.

Reply:
and I don't know if this makes any difference at all in practice, I don't have a feel for how performance-critical this is. :-)

Reply:
ok